lib/direct.ts

```diff
@@ -0,0 +1,112 @@
+import { KeyringPair } from "@polkadot/keyring/types";
+import ReconnectingWebSocket from "reconnecting-websocket";
+import { connect } from "./connect";
+import { createIdentity } from "./identity";
+import { newJWT } from "./jwt";
+import { challenge, sign } from "./sign";
+import { Address, Envelope, Request } from "./types/types_pb";
+import { waitReady } from '@polkadot/wasm-crypto';
+import { v4 as uuidv4 } from 'uuid';
+import { ApiPromise, WsProvider } from '@polkadot/api'
+export interface directClientInterface {
+    source: Address,
+    signer: KeyringPair,
+    twinId: number,
+    url: string,
+    responses: Map<any, any>
+}
+export async function getTwinId(address: string) {
+    const provider = new WsProvider("wss://tfchain.dev.grid.tf/ws")
+    const cl = await ApiPromise.create({ provider })
+    const twin = await cl.query.tfgridModule.twinIdByAccountID(address);
+    console.log(twin)
+    cl.disconnect();
+    return twin;
+
+}
+export async function newDirectClient(url: string, session: string, mnemonics: string, accountType: string) {
+    await waitReady();
+
+    // create identity of source
+    const identity = createIdentity(mnemonics, accountType);
+    const twinId = Number(await getTwinId(identity.address));
+    // create token from identity
+    const token = newJWT(identity, twinId, session)
+
+    // update url with token
+    url = `${url}?${token}`;
+
+    // create source from twin id and session string using generated proto types
+    const source = new Address();
+    source.setTwin(twinId);
+    source.setConnection(session);
+
+    // connect websocket
+
+    const responses = new Map<any, any>()
+
+    // create client with websocket connection
+    const client: directClientInterface = {
+        source: source,
+        signer: identity,
+        twinId: twinId,
+        url: url,
+        responses: responses
+    }
+
+    return client;
+
+}
+export async function createDirectClient(url: string, session: string, mnemonics: string, keyType: string) {
+    // create client
+    const client = await newDirectClient(url, session, mnemonics, keyType);
+
+    return client;
+
+}
+export function sendDirectRequest(client: directClientInterface, socket: ReconnectingWebSocket, requestCommand: string, requestData: any[], destinationTwinId: number) {
```

Comment:
does it a must to have the `requestData` as list? can't send an object (the use case is sending the deployment object to the node)?

Reply:
as protobuf types require string or Uint8Array any type that can be json.stringify then converted to a buffer will do